deepchem/models/torch_models/layers.py

```diff
@@ -6363,3 +6359,133 @@ def forward(self, inputs: List[np.ndarray]) -> torch.Tensor:
             deg_maxed.insert(0, self_atoms)
 
         return torch.concat(deg_maxed, 0)
+
+
+class GraphGather(nn.Module):
+    """A GraphGather layer pools node-level feature vectors to create a graph feature vector.
+
+    Many graph convolutional networks manipulate feature vectors per
+    graph-node. For a molecule for example, each node might represent an
+    atom, and the network would manipulate atomic feature vectors that
+    summarize the local chemistry of the atom. However, at the end of
+    the application, we will likely want to work with a molecule level
+    feature representation. The `GraphGather` layer creates a graph level
+    feature vector by combining all the node-level feature vectors.
+
+    One subtlety about this layer is that it depends on the
+    `batch_size`. This is done for internal implementation reasons. The
+    `GraphConv`, and `GraphPool` layers pool all nodes from all graphs
+    in a batch that's being processed. The `GraphGather` reassembles
+    these jumbled node feature vectors into per-graph feature vectors.
+
+    Example
+    --------
+    >>> import deepchem as dc
+    >>> import numpy as np
+    >>> import deepchem.models.torch_models.layers as torch_layers
+    >>> batch_size = 2
+    >>> raw_smiles = ['CCC', 'C']
+    >>> from rdkit import Chem
+    >>> mols = [Chem.MolFromSmiles(s) for s in raw_smiles]
+    >>> featurizer = dc.feat.graph_features.ConvMolFeaturizer()
+    >>> mols = featurizer.featurize(mols)
+    >>> multi_mol = dc.feat.mol_graphs.ConvMol.agglomerate_mols(mols)
+    >>> atom_features = multi_mol.get_atom_features().astype(np.float32)
+    >>> degree_slice = multi_mol.deg_slice
+    >>> membership = multi_mol.membership
+    >>> deg_adjs = multi_mol.get_deg_adjacency_lists()[1:]
+    >>> args = [atom_features, degree_slice, membership] + deg_adjs
+    >>> result = torch_layers.GraphGather(batch_size)(args)
+    >>> type(result)
+    <class 'torch.Tensor'>
+    >>> result.shape
+    torch.Size([2, 150])
+
+    References
+    ----------
+    .. [1] Duvenaud, David K., et al. "Convolutional networks on graphs for
```

Comment:
Can we change this to a specific name in [1] to something like [graph1] so that the reference can work sitewide